src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -199,15 +207,21 @@ class UserService(
         // validate all fields to be updated
         var errorReports = Seq[ErrorReport]()
         request.email.foreach(email => errorReports = errorReports ++ validateEmail(email, blockedEmailDomains))
+        request.azureB2CId.foreach(azureB2CId => errorReports = errorReports ++ validateAzureB2CId(azureB2CId))
+        errorReports = errorReports ++ request.isValid(user)
         if (errorReports.nonEmpty) {
           IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "invalid user update", errorReports)))
         } else { // apply all updates
-          var updatedUser = user
-          request.email.foreach { email =>
-            directoryDAO.updateUserEmail(userId, email, samRequestContext)
-            updatedUser = user.copy(email = email)
-          }
-          IO(Some(updatedUser))
+          val updatedUser = if (request.azureB2CId.isDefined || request.googleSubjectId.isDefined) {
+            directoryDAO.updateUser(user, request, samRequestContext)
+          } else IO(Option(user))
+
+          if (request.email.isDefined) {
+            updatedUser.map { u =>
+              directoryDAO.updateUserEmail(userId, request.email.get, samRequestContext)
```

Comment:
IDK if we should be updating the user email at this point. There's more that needs to happen, like putting the new email into the proxy group, updating the all users group, updating other systems, etc.